enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj

```diff
@@ -152,43 +112,51 @@
 
   [[serdes/descendants]] is recursively called on these entities and all their descendants, until the
   complete transitive closure of all descendants is found. This produces a set of `[\"ModelName\" id]` pairs, which
-  entities are then extracted the same way as [[extract-metabase]].
-Eg. if Dashboard B includes a Card A that is derived from a
-  Card C that's in an alien collection, warnings will be emitted for C, A and B, and all three will be excluded from the
-  serialized output."
-  [{:keys [targets] :as opts}]
+  entities are then returned as a reducible stream of serializable entity maps, with `:serdes/meta` keys.
+
+  Eg. if Dashboard B includes a Card A that is derived from a Card C that's in an alien collection, warnings will be
+  emitted for C, A and B, and all three will be excluded from the serialized output.
+
+  `opts` are passed down to [[serdes/extract-all]] for each model."
+  [{:keys [targets user-id] :as opts}]
   (log/tracef "Extracting subtrees with options: %s" (pr-str opts))
-  (let [targets  (->> targets
-                      (mapv (fn [[model-name id :as target]]
-                              (if (number? id)
-                                target
-                                [model-name (serdes/eid->id model-name id)]))))
-        analysis (escape-analysis targets)]
-    (if analysis
-      ;; If that is non-nil, emit the report.
-      (escape-report analysis)
-      ;; If it's nil, there are no errors, and we can proceed to do the dump.
-      ;; TODO This is not handled at all, but we should be able to exclude illegal data - and it should be
-      ;; contagious. Eg. a Dashboard with an illegal Card gets excluded too.
-      (let [nodes       (set/union
-                         (u/traverse targets #(serdes/ascendants (first %) (second %)))
-                         (u/traverse targets #(serdes/descendants (first %) (second %))))
-            models      (model-set opts)
-            ;; filter the selected models based on user options
-            by-model    (-> (group-by first nodes)
-                            (select-keys models)
-                            (update-vals #(set (map second %))))
-            extract-ids (fn [[model ids]]
-                          (serdes/extract-all model (merge opts {:where [:in :id ids]})))]
+  (let [inner-targets (if (seq targets)
+                        (mapv parse-target targets)
+                        (mapv vector (repeat "Collection") (collection-set-for-user user-id)))
+        nodes         (set/union
+                       (u/traverse inner-targets #(serdes/ascendants (first %) (second %)))
+                       (u/traverse inner-targets #(serdes/descendants (first %) (second %))))
+        reasons       (u/index-by #(take 2 %) #(first (nnext %)) nodes)
```

Comment:
I think using destructuring here would be good for self-documentation?

```suggestion
        reasons       (u/index-by #(subvec % 0 2)
                                  (fn [[_model _id maybe-more]] maybe-more)
                                  nodes)
```

or just `#(get % 2)`, which doesn't throw `IndexOutOfBoundsException`s like `nth`.